File: 1-easy/problems/0566-reshape-the-matrix/solution.cpp
class Solution {
public:
    vector<vector<int>> matrixReshape(vector<vector<int>>& mat, int r, int c) {
        int rows = mat.size();
        int cols = mat[0].size();
        if (rows * cols != r * c) {
            return mat;
        }

        vector<vector<int>> result(r, vector<int>(c));
        for (int index = 0; index < rows * cols; index++) {
            result[index / c][index % c] = mat[index / cols][index % cols];
        }
        return result;
    }
};

File: 1-easy/problems/0566-reshape-the-matrix/solution.ts
function matrixReshape(mat: number[][], r: number, c: number): number[][] {
    const rows = mat.length;
    const cols = mat[0].length;
    if (rows * cols !== r * c) {
        return mat;
    }

    const result = Array.from({ length: r }, () => new Array<number>(c));
    for (let index = 0; index < rows * cols; index++) {
        result[Math.floor(index / c)][index % c] = mat[Math.floor(index / cols)][index % cols];
    }
    return result;
}
